### ISSUE TITLE
Support for Cassandra in Kubernetes

### ISSUE THREAD

**Issue by bwmills** [thelastpickle/cassandra-medusa#24]
Greetings,

This looks like a great and much-needed project.

Saw the comment from this [TLP blog post](https://thelastpickle.com/blog/2019/11/05/cassandra-medusa-backup-tool-is-open-source.html) regarding compatibility with Cassandra running in Docker containers (e.g. a Kubernetes stateful set using persistent volumes for storage), and want to confirm that you do **not** recommend testing `cassandra-medusa` in Kubernetes at this time.

**Comment by adejanovski** [comment]
Hi @bwmills,

Thanks!

At the very least for restores, Medusa will open ssh connections to the nodes, which will clearly not work with k8s.
So there's indeed a little thinking and some work to do before we can make it work there.

**Comment by bwmills** [comment]
Thanks @adejanovski for the fast response. 

This is of course a +1 for Kubernetes support in the future. Have a great holiday.

**Comment by bwmills** [comment]
Actually @adejanovski if you have a moment to clarify - are you referring to the ephemeral nature of Kubernetes VMs (nodes) in general or to:

> Medusa will need to run without sudo as it will connect through ssh to all nodes in the cluster in order to perform remote operations.

Would it be possible to run `cassandra-medusa` as a cron job, where the pod deployed by the cron job uses the Google Cloud SDK ( `gcloud compute ssh ...` ) to perform remote operations?

I need to do some more reading about `cassandra-medusa` and so apologies if my question/idea is somehow not viable.

**Comment by jsanda** [comment]
Hey @bwmills 

I have spent a little bit of time thinking about this. When you do a cluster-wide restore,

```
$ medusa restore-cluster
```
medusa takes on the role of a control node or orchestrator. It does ssh into each machine and then run `medusa restore-node` (if I remember correctly). The `restore-node` command roughly does the following:

* download the backup
* stop Cassandra
* restore the backup
* start Cassandra

I would be very inclined to write a k8s operator for medusa. I would have a couple (maybe more) CRDs, something like `MedusaBackupJob` and `MedusaRestoreJob`. The CRDs would allow you to configure things like:

* access to the remote storage
* schedule for reoccurring backups
* the target cluster to backup
* the destination cluster to restore

I would not expose or use the `restore-cluster` command. The operator would facilitate the role of control node. Suppose we have a three-node cluster - C1, C2, C3 - that we want to restore. We create a `MedusaRestoreJob` object with the necessary configuration. The operator in turn creates a job that performs a restore on C1. After that job completes the operator creates a job to restore C2 and then C3. I am glossing over a lot of details, but I just want to provide a high level picture at the moment. These k8s jobs would run `medusa restore-node`, not `restore-cluster` so I don't think SSH would be a big concern. That is not to suggest that other changes are not necessary to make medusa work in k8s.

Dealing with the statefulsets, PVs, and PVCs gets interesting. The restore process involves stopping and starting the Cassandra nodes. To shut down C1 we simply kill the pod in which it is running; however, the StatefulSet controller will immediately recreate the pod. We will have to deal with this race condition. I did some research on this and I would delete the StatefulSet(s) without cascading the deletes so that all pods remain intact. The medusa operator is then free to kill a pod, perform whatever work is necessary, recreate the pod, and then recreate the StatefulSet object. The StatefulSet controller will discover the existing pods so the StatefulSet will return to the ready state very quickly.

Now onto the PVs and PVCs. When I was brainstorming this, I first though about restoring a single C* node and I assumed `ReadWriteOnce` access mode for the PV. When the pod is shutdown, we have to delete the existing PVC for the data volume. A new PVC is created for the `MedusaRestoreJob` so that the pod in which medusa is running can mount the data volume. We might also need another PV on which to download the backup. After the restore job finishes, we recreate the PVC for the C1 pod and then recreate the pod. Then we recreate the StatefulSet object.

I did a little bit of prototyping to check that these steps work, but I still need to spend time reviewing the StatefulSet controller code to make sure I understand the behavior with this sort of workflow.

I think the steps for a cluster-wide restore might actually be a bit less convoluted. We could potentially create new PVs and a new statefulset. We could then use label selectors to bind the new PVs to the new pods in the statefulset. We would then delete the existing statefulset and create a new one that uses the restore PVs.

The operator needs to know about the cluster topology as well. For instance, is the entire C* cluster running as a single statefulset? Or is there is a statefulset per rack? Or per DC? That's a whole other discussion :)

**Comment by bwmills** [comment]
@jsanda - this is incredibly helpful - thank you!

I have some thoughts on all this (especially dealing with PVs). I am off for the Thanksgiving holiday but will follow-up soon. Much appreciated.

**Comment by jsanda** [comment]
I am off for Thanksgiving as well. More than happy to collaborate on this. Enjoy the holidays :)

**Comment by gruaig** [comment]
Here is an error from a restore in a docker container.

> `[2019-11-28 00:06:08] INFO: Stopping Cassandra
Traceback (most recent call last):
  File "/usr/local/bin/medusa", line 8, in <module>
    sys.exit(cli())
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/click/decorators.py", line 64, in new_func
    return ctx.invoke(f, obj, *args[1:], **kwargs)
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/medusa/medusacli.py", line 180, in restore_node
    verify, set(keyspaces), set(tables), use_sstableloader)
  File "/usr/local/lib/python3.7/site-packages/medusa/restore_node.py", line 48, in restore_node
    keyspaces, tables)
  File "/usr/local/lib/python3.7/site-packages/medusa/restore_node.py", line 84, in restore_node_locally
    cassandra.shutdown()
  File "/usr/local/lib/python3.7/site-packages/medusa/cassandra_utils.py", line 408, in shutdown
    subprocess.check_output(self._stop_cmd)
  File "/usr/local/lib/python3.7/subprocess.py", line 411, in check_output
    **kwargs).stdout
  File "/usr/local/lib/python3.7/subprocess.py", line 488, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/usr/local/lib/python3.7/subprocess.py", line 800, in __init__
    restore_signals, start_new_session)
  File "/usr/local/lib/python3.7/subprocess.py", line 1551, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'sudo': 'sudo'
`

**Comment by jsanda** [comment]
@gruaig Medusa does unfortunately use `sudo`. You will need to include `sudo` in your image.

**Comment by giltohar** [comment]
@jsanda @bwmills is there any update on a medusa operator for cassandra on kubernetes?

**Comment by jsanda** [comment]
Hi @giltohar 

I have had some discussions with @adejanovski and @rzvoncek about next steps, but nothing concrete from me to date. I am however planning to start doing some of the prerequisite work this and happy to provide updates here.

Using medusa as a sidecar with a Cassandra k8s operator seems like the most obvious use case to me. 

@giltohar do you have/want use cases for a medusa operator on its own? If so, would you please share your thoughts?

**Comment by bwmills** [comment]
@giltohar - thanks for checking in. Though I've not had time to dive into this, the ability to run Medusa with Kubernetes still has great potential value to our team, and we remain very interested.

@jsanda - glad to hear you've kept some conversion going on this thread - I may have some time during the last week of May to do some actual testing.

**Comment by giltohar** [comment]
My team has decided to deploy the Bitnami package of Apache Cassandra on our Kubernetes cluster. We are preparing it for production use, and backup/restore is our last major prerequisite for production readiness.

For now, until a proper operator is ready, I am thinking of deploying a daemonset, which i can grant admin access, and having it run a script daily to create backups and send them to our backup storage.

**Comment by jsanda** [comment]
@adejanovski created #129 so that we can map out how to proceed. It is very likely that we will want a _service layer_ on top of medusa. This might be REST or gRPC for example. #88 was created a while back. I am going to start working on that.

I am not sure how feasible or practical it is to have a medusa operator on its own. Let me clarify some terminology to make things clear. Kubernetes runs controllers that manager things like deployments, statefulsets, pods, etc. An operator is one or more controllers plus one or more custom resource definitions with domain-specific logic built in. 

I definitely see the need for a controller, but I am not so sure about the need for an operator particularly when you consider restore scenarios. Whether restoring in-place or to a new cluster, there a lot of steps involved that are or should be performed by a k8s operator for Cassandra. Given this, I am more inclined to focus on a sidecar that can be used with a Cassandra operator.

If anyone has strong thoughts/opinions on a stand alone operator for medusa, please share. If there is way to perform necessary operations without reinventing work that is already done by Cassandra operators, then I am all for it.

**Comment by bwmills** [comment]
Hi @giltohar 

Good luck with your plans. I'm guessing that you and your team have done significant research and planning on production readiness, so the information I've provided below may not be relevant. I've included it on the off chance that some aspect may be of use to you or another reader.

**Overview**

We've been running a small Cassandra cluster as a Kubernetes `StatefulSet` of three replicas using persistent volumes for storage for several years now and it's been very solid. The Cassandra cluster runs in GKE and uses the `GoogleCloudSnitch`, 4 tokens, 3 racks (with a single node in each rack, where each rack correlates to a zone in our region), and a replication factor of 3.

**StatefulSet**

As we use Prometheus and Grafana for a portion of of monitoring services, the `StatefulSet` manifest includes the required annotations:

```
annotations:
  prometheus.io/port: "7090"
  prometheus.io/scrape: "true"
```

The manifest also uses some `podAntiAffinity` rules to ensure a balanced deployment across our thee zones (i.e. one Cassandra node per zone). 

```
  spec:
    affinity:
      podAntiAffinity:
        preferredDuringSchedulingIgnoredDuringExecution:
        - podAffinityTerm:
            labelSelector:
              matchExpressions:
              - key: app
                operator: In
                values:
                - cassandra
            topologyKey: failure-domain.beta.kubernetes.io/zone
          weight: 100
        - podAffinityTerm:
            labelSelector:
              matchExpressions:
              - key: app
                operator: In
                values:
                - cassandra
            topologyKey: kubernetes.io/hostname
          weight: 50
```

All our nodes are organized into node pools, and we have a dedicated pool for Cassandra with machine types that are appropriate to our platform. The node pool is configured with: 

**Kubernetes labels:** pool:cas
**Taints:** NoSchedule pool=cas 

and (to ensure Cassandra pods always run in that pool) the manifest includes the matching config:

```
    nodeSelector:
      pool: cas
    tolerations:
    - effect: NoSchedule
      key: pool
      operator: Equal
      value: cas
```

The `updateStrategy` in the manifest uses `OnDelete` (much preferred), and we have a storage class to ensure the nodes use SSD.

```
  updateStrategy:
    type: OnDelete
  volumeClaimTemplates:
  - metadata:
      name: datadir
    spec:
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 100Gi
      storageClassName: ssd
```

There are of course additional configs in the `StatefulSet` manifest that I haven't included here as this comment is getting long. They include the readiness and liveness probes, `ports`, `resources`, `volumes`, `volumeMounts`, `securityContext`, and so forth.

It may be worth mentioning a final section of the manifest we use to safely drain a Cassandra node when a pod is rolled or otherwise deleted.

```
  lifecycle:
    preStop:
      exec:
        command:
        - /bin/sh
        - -c
        - nodetool drain
```

**Service**

We use a `ClusterIP` (headless) service to expose the Cassandra service on internal IPs:

```
apiVersion: v1
kind: Service
metadata:
  labels:
    app: cassandra
  name: cassandra-svc
spec:
  clusterIP: None
  ports:
  - name: cql
    port: 9042
    protocol: TCP
    targetPort: 9042
  - name: prometheus-jmx
    port: 7090
    protocol: TCP
    targetPort: 7090
  - name: intra-node
    port: 7000
    protocol: TCP
    targetPort: 7000
  - name: tls-intra-node
    port: 7001
    protocol: TCP
    targetPort: 7001
  - name: jmx
    port: 7199
    protocol: TCP
    targetPort: 7199
  selector:
    app: cassandra
```

**PDB**

We use a simple `PodDisruptionBudget` as you would imagine:

```
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
metadata:
  name: cassandra-pdb
spec:
  minAvailable: 2
  selector:
    matchLabels:
      app: cassandra
```

**ConfigMap**

We use `ConfigMap` to manage a few of the key environment variables that are best handled outside of the `StatefulSet` manifest. These environment variables include two that relate to our current restore capabilities, which are indeed very simple. One variable is a boolean to enable/disable a restore, and the other points to the the tar file in Google Cloud Storage (GCS) where the backups reside. 

```
CASSANDRA_RESTORE: "false"
CASSANDRA_RESTORE_URL: "https://storage.googleapis.com/...redacted.../production-seed-0-04-06-2020.tgz"
```

This simplistic restore operation works, but is - as you can see - limited. This is why we're interested in `cassandra-medusa` and some of the granularity it provides.

**Images**

We manage the Cassandra Docker image ourselves. The base image is `ubuntu-slim`, then another image: `java-ubuntu-slim` runs `OpenJDK 8u222`, then the third image builds Cassandra `3.11.2`. 

This last image also installs the [jmx_exporter](https://github.com/prometheus/jmx_exporter) and runs it as a Java Agent in the container. Metrics are exported in the Prometheus data model, scraped by a Prometheus instance running in our GKE cluster, and we use Grafana as a frontend as you might expect. There is a second Cassandra `ConfigMap` that defines our metrics of interest - this manifest effectively filters out the metrics we don't need using the syntax defined by the exporter linked above. 

A long comment here to be sure, and most of it is not relevant to the original topic, but I've found that - outside of the Cassandra mailing list - there seems to be less information about running Cassandra with Kubernetes than I would like. Hopefully something here is of use to you and other readers.

**Comment by jsanda** [comment]
>A long comment here to be sure, and most of it is not relevant to the original topic

@bwmills I have to respectfully disagree :) The info is super helpful. Thank you for sharing all the details.

>If CASSANDRA_RESTORE is true when a Cassandra pod is restarted - let's say the pod is cassandra-0 - data in the PV attached to the node (the GKE VM instance where that pod is running) is overwritten by the untarred backup file

Do you have code in your container's entrypoint script that copies the backup from GCS?

Let me give a quick brain dump of what I have been thinking about. We put a REST (or gRPC) service layer on top of medusa. Medusa runs as a _sidecar_ container in the same pod as the Cassandra container. For scheduled backups, I was thinking along the lines of  what @bwmills described. Have a CronJob that will call each medusa sidecar to do the backup. Alternatively you could do a separate CronJob per Cassandra pod, but I am not sure how that would scale for large C* clusters. 

In the case of a restore, we want the Cassandra container to start after the medusa container. Unfortunately, there currently is no way to specify startup dependencies between containers. See [this ticket](https://github.com/kubernetes/kubernetes/issues/65502) for more info. We would need to [init container](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/) to ensure medusa runs before Cassandra starts.

>I've found that - outside of the Cassandra mailing list - there seems to be less information about running Cassandra with Kubernetes than I would like. 

Agreed. I want to point out some relevant blog posts I have written:

* [Creating a Cassandra Cluster with Cass Operator](https://thelastpickle.com/blog/2020/04/15/deploy-cassandra-with-k8s-operators.html)
* [Deploying Cassandra with CassKop](https://itnext.io/deploying-cassandra-in-kubernetes-with-casskop-98956edfe32b)
* [Create Affinity between Cassandra and Kubernetes](https://medium.com/@john.sanda/create-affinity-between-cassandra-and-kubernetes-5b2f72d23293)

**Comment by rrusmana** [comment]
Hi @bwmills I wanted to get a better idea of your CronJob for backup. I'm assuming you are rolling your own scripts to do the snapshot and upload to gcs (i.e. not using medusa)? Can you share the script? I have been evaluating https://github.com/maorfr/cain to do this and it uses a similar approach but I have been encountering a problem during uploads (uploads hanging without error). Have you encountered something similar in your experience?

**Comment by bwmills** [comment]
@jsanda - you are too kind, thank you - I'm glad those details look to be of some use.

> Do you have code in your container's entrypoint script that copies the backup from GCS?

Yes, our Dockerfile finished with this: CMD instruction:

**Comment by jsanda** [comment]
I want to provide a quick status update on what I have been doing and thinking about.

I have started working on creating a gRPC service for Medusa. This is being tracked under #137.
The goal is to have the gRPC service running in a sidecar container to handle backup requests. 

I think we will need a different Dockerfile image than what is currently in the project for the sidecar.

Those requests will come from a k8s controller for Medusa. I will probably work on that in a separate repo as it will be Go code and more k8s-specific.

We could create two CRDs - `Backup` and `BackupJob`. The former would specify that a backup should be created. It could use label selectors for instance to determine which Cassandra pods should be backed up. The `BackupJob` would CRD would facilitate scheduled backups.

For restores, we can use an init container in addition to the sidecar container. Init containers run to completion before the main containers in the pod start. To perform a restore then, we can delete the Cassandra pod (or just the container I suppose). The StatefulSet controller will create a new pod. Before starting the Cassandra container, the Medusa init container will run. 

Things might be a bit different when integrating with a k8s operator for Cassandra. This approach though provides a path forward without requiring a Cassandra operator.

**Comment by JensRantil** [comment]
Looks like it's supported now that https://github.com/k8ssandra/k8ssandra ships with Medusa?

**Comment by bwmills** [comment]
Many thanks @JensRantil - I had not seen that!

**Comment by adejanovski** [comment]
Hey folks, 

@jsanda will correct me if I'm wrong but we first need to merge #223 to make Medusa work in k8s. You'll probably need to check how Medusa support was implemented in k8ssandra as well if you want to replicate it in your environment.

**Comment by jsanda** [comment]
k8ssandra is currently using a patched version of Medusa with k8s images built from the [grpc](https://github.com/thelastpickle/cassandra-medusa/tree/grpc) branch.

@emerkle826 is working on the outstanding changes needed to get #223 merged. 

Check out the k8ssandra [docs](https://k8ssandra.io/docs/topics/restore-a-backup/) for an example of backup/restore. Things are still early and still lots of work to do. I will be adding issues in [medusa-operator](https://github.com/k8ssandra/medusa-operator) over the next few days. Stay tuned!

**Comment by emerkle826** [comment]
> @emerkle826 is working on the outstanding changes needed to get #223 merged.

I am working this and trying to get it ready for merge as I type.

**Comment by jsanda** [comment]
>You'll probably need to check how Medusa support was implemented in k8ssandra as well if you want to replicate it in your environment.

Also on my TODO list is a blog post to describe this :)

**Comment by bwmills** [comment]
Thanks everyone for these updates 😄

**Comment by arodrime** [comment]
@jsanda should we now close this ticket? 

Is K8ssandra still on a patched version of medusa?
Does Medusa now work with other K8s solutions as well?

**Comment by jsanda** [comment]
>should we now close this ticket?

Yes we can close this ticket.

>Is K8ssandra still on a patched version of medusa?

No. We are now using images built from master!

>Does Medusa now work with other K8s solutions as well?

It could but would require a good bit of work. It would require the same amount of effort that is done in k8ssandra which includes [medusa-operator](https://github.com/k8ssandra/medusa-operator). One other thing I should note is that things are designed specifically to work with cass-operator.  If anyone would like further discussion/docs around this, then let's address in a separate ticket.
